[feature/pasteboard] Pasteboard Actions

hosy opened the pull request:
## Description
This PR implements new actions, regarding the system and app pasteboard:
- Copy
- Paste (Import)
- Cut

Inside the app it is possible to cut and copy items from one path to a new path.
Outside the app, it is possible to provide items to other apps and the system.

## Related Issue

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%2011.6.1/Pasteboard%20options.md

Reports:

- [x] (1) Copy with multiselection pastes only one item https://github.com/owncloud/ios-app/pull/514#issuecomment-880461961 [FIXED]
- [x] (2) Notification when folder is copied to clipboard https://github.com/owncloud/ios-app/pull/514#issuecomment-880499884 [FIXED]
- [x] (3) Copy to external app https://github.com/owncloud/ios-app/pull/514#issuecomment-880508833 [NOT SUPPORTED]
- [X] (4) Bad file name copy/pasting from/to Files App https://github.com/owncloud/ios-app/pull/514#issuecomment-880514041  [FIXED]
- [X] (5) Copy/paste text from external apps https://github.com/owncloud/ios-app/pull/514#issuecomment-880541603 [DONE]
- [X] (6) Cut option not working https://github.com/owncloud/ios-app/pull/514#issuecomment-880569992 [EXPECTED]
- [X] (7) Limited storage for items in clipboard? https://github.com/owncloud/ios-app/pull/514#issuecomment-883356739 [TO RESEARCH]

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=514) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=514) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: hosy<br/>:white_check_mark: ownclouders<br/>:white_check_mark: mneuwert<br/>:x: felix-schwarz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=514) it.</sub>

mneuwert commented on ownCloud/Client/Actions/Actions+Extensions/CopyAction.swift:
```diff
@@ -69,11 +104,54 @@ class CopyAction : Action {
+			// General system-wide Pasteboard
+			if item.type == .collection {
+				let pasteboard = UIPasteboard.general
+				tabBarController.pasteboardChangedCounter = pasteboard.changeCount
```
@hosy curious, why `UITabController` instance is used to keep track of paste board change count?

hosy replied:
I moved that code to the Key-Value store.

mneuwert commented on ownCloud/Client/Actions/Actions+Extensions/CopyAction.swift:
```diff
@@ -69,11 +104,54 @@ class CopyAction : Action {
+				}
+
+				core.downloadItem(item, options: [ .returnImmediatelyIfOfflineOrUnavailable : true ], resultHandler: { (error, core, item, _) in
+					if error == nil {
```
@hosy reduce nesting levels with guards?

hosy replied:
done!

mneuwert commented on ownCloud/Client/Actions/Actions+Extensions/ImportPasteboardAction.swift:
```diff
@@ -0,0 +1,125 @@
+		let pasteboard = UIPasteboard.general
+
+		// Determine, if the internal pasteboard is the current item and use it
+		if pasteboard.changeCount == tabBarController.pasteboardChangedCounter {
```
@hosy reduce nesting with guards?

hosy replied:
done!

mneuwert commented on ownCloud/Client/Actions/Actions+Extensions/ImportPasteboardAction.swift:
```diff
@@ -0,0 +1,125 @@
+												OCCoreOption.automaticConflictResolutionNameStyle : OCCoreDuplicateNameStyle.bracketed.rawValue
+							],
+											 placeholderCompletionHandler: { (error, item) in
+												if error != nil {
```
@hosy errors returned by SDK are replaced by internal error. Wouldn't it lead to masking of the error in some cases? Like you paste in some file and there is no quota available but the actual error won't be visible in the app?

hosy replied:
You are right, I replaced it with the SDK errors

jesmrec commented:
### QA

Clipboard

- [X] Copy one file / Paste in another location -> correctly pasted
- [X] Cut one file / Paste in another location -> correctly pasted and removed from original location
- [X] Copy one folder / Paste in another location -> correctly pasted
- [X] Cut one folder / Paste in another location -> correctly pasted and removed from original location
- [ ] Copy many files / Paste in another location -> correctly pasted
- [ ] Cut many files / Paste in another location -> correctly pasted and removed from original location
- [ ] Copy many folders / Paste in another location -> correctly pasted
- [ ] Cut many folders / Paste in another location -> correctly pasted and removed from original location

Choose directory

- [X] Copy one file / Paste in another location -> correctly pasted
- [X] Copy one folder / Paste in another location -> correctly pasted
- [ ] Copy many files / Paste in another location -> correctly pasted
- [ ] Copy many folders / Paste in another location -> correctly pasted

jesmrec commented:
### (1)

1. Select many files
2. Select option `Copy` 
3. Select `Copy to Clipboard`
4. Go to another location and select `Paste`

Current: only one item is pasted
Expected: all copied items are pasted

`6e9c9755`
iPhoneXR v14.6

jesmrec commented:
### (2)

1. Select a folder
2. Select `Copy` -> `Copy to Clipboard`

Current: Notification `Item was copied to the clipboard`  is missing.

Expected: Notification `Item was copied to the clipboard`  appears, like files do.

`6e9c9755`
iPhoneXR v14.6

jesmrec commented:
### (3)

1. Select a folder in the app
2. Select `Copy` -> `Copy to Clipboard`
3. Open Files App
4. Paste the copied folder somewhere

Current: not pasted. 
Expected: folder and its content is pasted

`6e9c9755 `
iPhoneXR v14.6

jesmrec commented:
### (4)

one question: 

Copying from oC to other location, f.ex., Files App, changes the item name. You can reproduce by following the steps from (3) with a file 

Copy in oC a gif file called `home.gif` -> Pasted in Files App with name `GIF image`
Copy in oC a pdf file called `manual.pdf` > Pasted in Files app with name `PDF document` 

is this a matter of oC or Files App?

If we do the same actions in the wrong way round, the effect is the same but the name slightly different

Copy in Files App a gif file called `home.gif` -> Pasted in oC app with name `com.compuserve.gif.gif`
Copy in Files App  a pdf file called `manual.pdf` > Pasted in oC app with name `com.adobe.pdf.pdf` 

is this something fixable?

jesmrec commented:
### (5)

question: is the following scenario in the scope of the current feature?:

1. Open Safari browser and open any website
2. Copy some text
3. Paste into oC

Current: it pastes several files: html, and rtf's but not the text

I know that the correct way to do it is by sharing... just want to know if this is expected here or not.

jesmrec commented:
### (6)

Option `Cut` works only inside the oC app, is that expected?

1. Open oC app, select a file and option `Cut`
2. Open Files App and paste in any location

Current: item not pasted
Expected: ?

`6e9c9755`
iPhoneXR v14.6

michaelstingl commented:
> ### (6)
> Option `Cut` works only inside the oC app, is that expected?

„Cut“ to another storage should be an implicit copy I‘d say.

jesmrec commented:
>> ### (6)
>> Option `Cut` works only inside the oC app, is that expected?

>„Cut“ to another storage should be an implicit copy I‘d say.

i've slightly modified the definition of (6): `Cut` only works in the same oC account. 

If cut content is pasted in another provider -> not pasted
If cut content is pasted in another oC account inside the app -> not pasted

hosy commented:
@jesmrec I fixed all QA findings. You can give it a new try.

- (1) multiple items can be copied and pastes 
- (2) "Copy" notification was fixed and shows the number of total copies items
- (3) copy a folder only works inside the oC app (folders cannot be pasted into Files.app or third party apps)
- (4) the pasteboard does not preserve the file name
- (5) pasting text (e.g. from Safari) results in more files (because the pasteboard provides different output formats. This may be solved in a further development step. But currently the user have different options, which format the user wants to keep)
- (6) cut an item is only possible within the account

jesmrec commented:
(1), (2) -> fixed
(3), (4) -> out of scope
(5) -> to improve in a separate issue?
(6) -> If `Cut` only works in the same account, is there any difference with the  `Move` option?

hosy commented:
> (1), (2) -> fixed
> (3), (4) -> out of scope
> (5) -> to improve in a separate issue?

Yes, we should have a look how it works for us and which ideas we have to solve this problem. But yes, this should be improved in a separate issue.

> (6) -> If `Cut` only works in the same account, is there any difference with the `Move` option?
You are right, it is like the `Move` option, but without the directory picker.

hosy commented:
@jesmrec I think I found a solution to preserve the file name. I will commit an update soon.

jesmrec commented:
about (3)

> (3) copy a folder only works inside the oC app (folders cannot be pasted into Files.app or third party apps)

Copying a folder from one oC account to another oC account inside the app. Should this work?

now, it doesn't

jesmrec commented:
another question: is there any space/storage restriction in the clipboard? copying many files (or less bigger files), they are not all pasted.

jesmrec commented:
After the 2nd round, pending stuff:

- Checking if pasted items could preserve the original name

- Folders (not) pasted to different account in the same app.

- Question about the size of the copied content

Some improvements that could be addressed in separate issues, as new iterations:

- Pasting text from clipboard: selecting format to paste the text ([report (5)](https://github.com/owncloud/ios-app/pull/514#issuecomment-880541603) )

- Add a spinner/waiter while the items are copied: if many items are selected to be copied to clipboard, some time runs till everything is copied and the notification appears. This could be confusing, so user does not know if the actions was correctly performed

- `Cut` option is only available for one item. Adding it to multiselection list

hosy commented:
@jesmrec I refactored the pasteboard code and you can give it a new try. 

> After the 2nd round, pending stuff:
> 
> * Checking if pasted items could preserve the original name
Items now preserve the original filename

> * Folders (not) pasted to different account in the same app.
At the moment folders cannot be pasted/cut between accounts. This needs more work but should be addressed in a new issue.

> * Question about the size of the copied content
Good question, this needs further research.

> 
> Some improvements that could be addressed in separate issues, as new iterations:
> 
> * Pasting text from clipboard: selecting format to paste the text ([report (5)](https://github.com/owncloud/ios-app/pull/514#issuecomment-880541603) )
Pasting text from clipboard now works as expected (only one file will be added)

> * Add a spinner/waiter while the items are copied: if many items are selected to be copied to clipboard, some time runs till everything is copied and the notification appears. This could be confusing, so user does not know if the action was correctly performed
I added a progress UI, when items will be copied.

> * `Cut` option is only available for one item. Adding it to multiselection list

`Cut` option is now available for multiselection list

jesmrec commented:
Feature is approved. 

But, it is approved under known limitations:

1) It is not posible to...

-  ... copy folders between accounts
-  ... copy folders to external provider
-  ... execute `Cut` action between accounts
-  ... execute `Cut` action against an external provider

2) Clipboard size to copy needs extra research as commented [here](https://github.com/owncloud/ios-app/pull/514#issuecomment-884181977). If the content copied inside the clipboard is great (pretty difficult to quantify, it depends on the device and its available storage), maybe the actions will not work fine, or the app gets stucked. 

Test plan with all test cases designed and executed in the top message

I will create issues for the known limitations.

hosy commented:
@jesmrec answering your questions:

> Feature is approved. 
> 
> 

Thanks

> 
> But, it is approved under known limitations:
> 
> 
> 
> 1) It is not posible to...
> 
> 
> 
> -  ... copy folders between accounts

Yes, this is at the moment not supported by the SDK. We can create a new issue for that problem.
I can add an alert view, when a user copied a folder and want to paste it to another account?

> 
> -  ... copy folders to external provider
> 

This is not supported by the iOS system at the moment.

> -  ... execute `Cut` action between accounts

This is working for files (expect for folders).

> 
> -  ... execute `Cut` action against an external provider
> 


This is not supported by the iOS system at the moment and also makes no sense in my opinion.

> 
> 
> 2) Clipboard size to copy needs extra research as commented [here](https://github.com/owncloud/ios-app/pull/514#issuecomment-884181977). If the content copied inside the clipboard is great (pretty difficult to quantify, it depends on the device and its available storage), maybe the actions will not work fine, or the app gets stucked. 
> 

Yes, needs research.

> 
> 
> Test plan with all test cases designed and executed in the top message
> 
> 
> 
> I will create issues for the known limitations.
> 
>
